### PR TITLE
MM-47750: Adds PostAcknowledgements table and apis

### PR DESCRIPTION
**Pull request by koox00**
#### Summary

- Adds post acknowledgement api/app/store methods to be able to save and
delete post acknowledgements by users.
- Adds wesbsocket events for acknowledgement created/deleted
- Returns post acknowledgements in the post's metadata

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-47750

#### Release Note

```release-note
Added new API endpoints POST /api/v4/users/:user_id/posts/:post_id/ack, DELETE /api/v4/users/:user_id/posts/:post_id/ack
Added new websocket events 'post_acknowledgement_added', 'post_acknowledgement_removed'.
Added a new Database table 'PostAcknowledgements'.
Added 'acknowledgements' field in the post's metadata.
```

### CODE REVIEW

**Comment by Rishselva** [comment]
Hello @ashishbhate & @vish9812! Please review this ticket as soon as you can, thank you :)

File: app/post_acknowledgements.go
Comment (Willyfrog) [inline app/post_acknowledgements.go:79]: is the 5 min configurable or is it supposed to be hardcoded?

File: store/sqlstore/post_acknowledgements_store.go
Comment (Willyfrog) [inline store/sqlstore/post_acknowledgements_store.go:128]: should we have a max number of posts to get?

**Review by Willyfrog — APPROVED** [review]
Added a couple of comments, but non blocking

File: api4/post.go
Comment (ashishbhate) [inline api4/post.go:990]: If we're getting "user_id" from the params, we should probably do a `RequireUserId()` on line 986.

I'm wondering if this is inconsistent with the `acknowledgePost` app method. There we get the "userID" from the request body, but here we get it from the query parameters.

Comment (koox00) [reply api4/post.go:990]: @ashishbhate, I was about to change that atm,

Would the most "restful" way be:
 
POST /ack (REQUEST {body: user_id})
DELETE /ack/{user_id}

File: app/post_acknowledgements.go
Comment (ashishbhate) [inline app/post_acknowledgements.go:30]: If a channel with a post requiring acknowledgement is archived, then the user won't be able to acknowledge the post, does that mean the user keep getting notifications?

Comment (koox00) [reply app/post_acknowledgements.go:30]: no, the persistent notifications entries for the channel will be deleted.
cc @vish9812

File: app/post_acknowledgements.go
Comment (lieut-data) [inline app/post_acknowledgements.go:119]: Can we pre-allocate this map based on the size of `acknowledgements`?

File: app/post_acknowledgements.go
Comment (lieut-data) [inline app/post_acknowledgements.go:124]: Can we eliminate this check and just rely on the `nil` handling semantics of `append`?

File: store/sqlstore/post_acknowledgements_store.go
Comment (lieut-data) [inline store/sqlstore/post_acknowledgements_store.go:30]: nit: could we order these to "match" the table? So `PostId`, `UserId`, and hten `AcknowledgedAt`.

File: store/sqlstore/post_acknowledgements_store.go
Comment (lieut-data) [inline store/sqlstore/post_acknowledgements_store.go:35]:
```suggestion
```

File: store/sqlstore/post_acknowledgements_store.go
Comment (lieut-data) [inline store/sqlstore/post_acknowledgements_store.go:23]: See earlier comment, but for example, should the parameters here be `postID, userID` for consistency?

File: db/migrations/mysql/000098_create_post_acknowledgements.up.sql
Comment (lieut-data) [inline db/migrations/mysql/000098_create_post_acknowledgements.up.sql:3]: nit: in a couple of places (e.g. parameters, model definiiton) I read the order `UserId, PostId` and in other cases, like here, I see `PostId, UserId`. I think this order makes sense -- should we be consistent everywhere?

File: store/sqlstore/post_acknowledgements_store.go
Comment (lieut-data) [inline store/sqlstore/post_acknowledgements_store.go:74]:
```suggestion
```

File: store/sqlstore/post_acknowledgements_store.go
Comment (lieut-data) [inline store/sqlstore/post_acknowledgements_store.go:80]:
```suggestion
```

File: store/sqlstore/post_acknowledgements_store.go
Comment (lieut-data) [inline store/sqlstore/post_acknowledgements_store.go:79]: Could we use `ExecBuilder` here and avoid the manual `query.ToSql()`?

File: store/sqlstore/post_acknowledgements_store.go
Comment (lieut-data) [inline store/sqlstore/post_acknowledgements_store.go:91]: This feels semantically equivalent to having no record at all, given we don't have something like a `CreatedAt` column. Should we just delete the row altogether?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Nice work, @koox00! A handful of comments.

File: app/post_acknowledgements.go
Comment (koox00) [reply app/post_acknowledgements.go:89]: To avoid getting the ack from the store a second time, but yes I can change to have the real value returned to the client.

File: app/post_acknowledgements.go
Comment (koox00) [reply app/post_acknowledgements.go:79]: It's supposed to be hardcoded.
cc @matthewbirtch

File: app/post_acknowledgements.go
Comment (koox00) [reply app/post_acknowledgements.go:119]: `acknowledgmentsMap` is actually acknowledgments by post and not all posts have acknowledgements

e.g let's say we have just 10 acks all for one post but we ask acks for 200 posts, the map would have one key (the post id) with a slice of 10 acks in that key.

```go
map[string]*[]PostAcknowledgements{
   [post1.Id]: [10]PostAcknoledgements,
}
```

File: app/post_acknowledgements.go
Comment (koox00) [reply app/post_acknowledgements.go:124]: I'm not sure how to do that. Don't we need to initialize the slice for each post?

File: app/post_acknowledgements.go
Comment (koox00) [reply app/post_acknowledgements.go:79]: @matthewbirtch oh sorry this one is the time the user has to un-acknowledge a post, not for persistent notifications

File: app/post_acknowledgements.go
Comment (lieut-data) [reply app/post_acknowledgements.go:119]: Aha, I misread that! Thank you for clarifying :)

File: app/post_acknowledgements.go
Comment (koox00) [reply app/post_acknowledgements.go:89]: @lieut-data now that I think of it, since we are deleting there is no way to get the deleted ack back, how can we avoid the race?

File: store/storetest/store.go
Comment (koox00) [reply store/storetest/store.go:112]: What... I don't think I've changed those 😮

File: app/post_acknowledgements.go
Comment (lieut-data) [reply app/post_acknowledgements.go:89]: I haven't seen the client-side code, but thinking out loud: since the ack's primary key is deterministic, I imagine that a `POST /ack` followed by a `DELETE /ack` followed by a `POST /ack` might result in some clients getting `ack, ack, delete` (e.g. out of order websockets). Since the only unique thing about each is the `AcknowledgedAt`, the client can just ignore anything with value strictly less than the currently known timestamp, keeping the second `ack` in this case and ignoring the `delete` to avoid the race.

File: app/post_acknowledgements.go
Comment (lieut-data) [reply app/post_acknowledgements.go:118]: Assuming I'm reading the code correctly (and I've already done that wrong once so far ;P), `acknowledgementsMap[ack.PostId]` starts out `nil`. When given a nil value (aka the empty value for a slice, itself subtly different than an empty slice), `append` simply allocates a new slice and returns that, making it safe, for example, to do this:

```go
var list []string
list = append(list, "first")
list = append(list, "second")
```

So the code in this case could just read:
```go
for _, ack := range acknowledgements {
    acknowledgementsMap[ack.PostId] = append(acknowledgementsMap[ack.PostId], ack)
}
```

Note that the map itself has to be initialized: the `nil` value for a map is safe to read from, but not safe to write to.

Comment (koox00) [reply app/post_acknowledgements.go:118]: oh I see, didn't know about that, thanks!

File: app/post_acknowledgements.go
Comment (koox00) [reply app/post_acknowledgements.go:88]: oh I see, ok I'll add the necessary changes in the webapp PR as well

File: app/post_acknowledgements.go
Comment (matthewbirtch) [reply app/post_acknowledgements.go:79]: Oh, my mistake, sorry it was hard to tell what this was referring to.

Yes, this wasn't something we identified as configurable so hardcoded I guess is the way to go

File: api4/api.go
Comment (lieut-data) [inline api4/api.go:59]: nit: I'm surprised to see a route at this level that's technically a descendant of the existing `Post` route. (In that I conceptually think of the `Post` route as defining everything underneath that path.)

I don't think there's anything strictly wrong with this, but just calling it out.

File: api4/post.go
Comment (lieut-data) [inline api4/post.go:43]: To confirm, is it intentional to have the `user_id` in the path for `DELETE` but not for `POST`?

File: api4/post.go
Comment (lieut-data) [inline api4/post.go:1000]: 0/5, but I'd love to be able to use this API without having to always specify myself as a user. Just to confirm that we don't want to make the `user_id` an optional querystring parameter?

(I often would type things like `http POST :8065/api/v4/posts/<post_id/ack` at the CLI.)

**Review by lieut-data — APPROVED** [review]
Thanks @koox00!

**Comment by koox00** [comment]
@jgilliam17 oh i realized i have different branch names the prs wont sync.

**Comment by jgilliam17** [comment]
@koox00 let me know when you have everything set. Thanks 🙂

File: api4/post.go
Comment (koox00) [reply api4/post.go:1000]: if we change the endpoint to include user_id in the url and be able to use `/user_id` or `/me`, would that work?

**Comment by koox00** [comment]
@jgilliam17 it that it won't create the server and I don't know why...

**Comment by jgilliam17** [comment]
@koox00 I've sent you a Split.io request to enable PostPriority, here on the test server created above

File: store/sqlstore/post_acknowledgements_store.go
Comment (ashishbhate) [reply store/sqlstore/post_acknowledgements_store.go:83]: Thinking ahead to how our mobile app syncs data (the mobile app asks for data that has changed since a given timestamp), I think having an `updateAt` column is the way to go. It would also simplify the possible race discussion in the app layer. 

If we just delete the row altogether, then then the mobile app will have to query every known Post Acknowledgement to check if it still exists.

**Review by ashishbhate — APPROVED** [review]
LGTM, just a two comments inline:

1. https://github.com/mattermost/mattermost-server/pull/21689#discussion_r1030222037 - is an implementation detail, so we add it later

2. https://github.com/mattermost/mattermost-server/pull/21689#discussion_r1030187406 - needs resolving now because its a API contract which becomes difficult to change later. I'll defer to you on what the API route should be

File: api4/api.go
Comment (lieut-data) [reply api4/api.go]: Ooh, yes! The best suggestion so far :)

**Comment by jgilliam17** [comment]
Thanks @koox00 
Here are few thing I noticed on my first pass
1. Help text theme not respected
Screen Shot 2022-11-23 at 11 21 01 AM

2. Long name adds scroll bar to acknowledged user list , we should truncate to avoid this
Screen Shot 2022-11-23 at 11 20 00 AM

3. Difference in icon color fill self-acknowledged vs. other user acknowledging. Ticket samples have solid green icon, maybe this was discussed in the PR and i missed it. Is it expected?
Screen Shot 2022-11-23 at 2 14 25 PM

4. What is the time limit to add acknowledgment/remove it? Should I be able to remove it? I didn't see anything in the ticket, but I can see it on the test server, older checks can't be undone but recent ones can.

**Comment by koox00** [comment]
Thank you @jgilliam17

> Thanks @koox00 Here are few thing I noticed on my first pass
> 
>     1. Help text theme not respected

This has been addressed, I wonder if the server doesn't have the latest webapp changes, let check on this.

>     2. Long name adds scroll bar to acknowledged user list , we should truncate to avoid this

oh, thanks, will truncate the names

>     3. Difference in icon color fill self-acknowledged vs. other user acknowledging. Ticket samples have solid green icon, maybe this was discussed in the PR and i missed it. Is it expected?

I never ask @matthewbirtch on this. As a user though I need to know if I had acknowledged a post or not, without scrolling through the list. 

>     4. What is the time limit to add acknowledgment/remove it? Should I be able to remove it? I didn't see anything in the ticket, but I can see it on the test server, older checks can't be undone but recent ones can.

It is 5min

**Comment by esethna** [comment]
@jgilliam17 I think (3) and (4) are expected. I agree there needs to be a way to differentiate when i personally have acked a post

**Comment by koox00** [comment]
> Also, @koox00 I don't think I was tagged as a reviewer on this PR, should I do a full UX review?

@matthewbirtch yes please :)

**Comment by koox00** [comment]
>  5. Once re-do timer expires on acknowledged post , tooltip fails to open on hover over check icon

@jgilliam17, hm weird, it opens for me
Screenshot 2022-11-23 at 9 51 29 PM

**Comment by matthewbirtch** [comment]
> > > ```
> > > 3. Difference in icon color fill self-acknowledged vs. other user acknowledging. Ticket samples have solid green icon, maybe this was discussed in the PR and i missed it. Is it expected?
> > > ```
> > 
> > 
> > I never ask @matthewbirtch on this. As a user though I need to know if I had acknowledged a post or not, without scrolling through the list.
> 
> 
> Actually, I don't know if I accounted for this in the design. I like the idea of the visual difference though. Let me take a look at it and I'll report back

Just confirming that I'm good with your approach for this @koox00

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
A few things I noticed in my testing beyond what's already been called out:

1. The divider line between ack and reactions should be using `--center-channel-color` as the theme variable NOT `--center-channel-text`. In dark themes, this divider is entirely hidden right now. (FYI this variable is going to be removed from the code base to avoid this in the future). 

2. When the ack button is in its default state, there is no divider between reactions and the ack button. It should still be there.
    
    image

3. Also, on the default acknowledge button, can we add 2px more of padding on the right?

4. Are we able to animate the width of the acknowledgment button when you click it and it shrinks to the acknowledged state? It looks like a CSS transition is applied, but doesn't seem to animate the width. And can we make it `ease-in-out` for the easing function? You may need to apply max-widths to both states to make that work.

3. The spacing below the message text and the acknowledgement button is inconsistent. When there are also reactions on the message, the spacing looks right, but when it is only showing the ack button, the spacing is too tight. This is also true for the space that displays _below_ the acknowledgment button and reactions.
    
    image

6. When I hover the ack button that I have already acknowledged, I get a disabled cursor. This makes sense if the ack undo time has passed, but it prevents me from seeing the popover of the list of people who have acknowledged. We should still allow the popover to display and we might want to consider displaying a reason for why I can no longer change my acknowledgement.

7. For the popover of the list of users who have acknowledged, we need to add a border on the popover container: 
 
    ```border: solid 1px rgba(var(--center-channel-color-rgb), 0.16);``` 

8. The popover for message priority also needs the same border as it gets lost in dark themes.

    ```border: solid 1px rgba(var(--center-channel-color-rgb), 0.16);``` 

    image

8. The ack buttons don't seem to display in the threads view or the RHS thread view. They should be showing in both those contexts as well

9. For the request acknowledgement toggle button that shows in the popover, I assume you're just using what is currently in the product. However, the toggle button style is to be updated. I don't think this has been done in the webapp anywhere yet though that I can think of. For now, at the very least, we should update the active color to use the `--button-bg` color variable.

10. Also, we should make the entire row clickable to turn acknowledgment on/off just like we do for the priority level. For that, we should also use the same hover state on the entire background of the row that we use on the priority levels.

**Comment by koox00** [comment]
@matthewbirtch, @esethna 


> The ack buttons don't seem to display in the threads view or the RHS thread view. They should be showing in both those contexts as well
> 
> For the request acknowledgement toggle button that shows in the popover, I assume you're just using what is currently in the product. However, the toggle button style is to be updated. I don't think this has been done in the webapp anywhere yet though that I can think of. For now, at the very least, we should update the active color to use the --button-bg color variable.
> 
> Also, we should make the entire row clickable to turn acknowledgment on/off just like we do for the priority level. For that, we should also use the same hover state on the entire background of the row that we use on the priority levels.

can we have the above as bug tickets to do the following days?

**Comment by esethna** [comment]
@koox00 yes we can do that!

**Comment by koox00** [comment]
@lieut-data can you please take a look at the load test results.

PS: these are created without having Acknowledgements in the table, but with having 500K of urgent posts, so essentially the same tests as the other PR with the urgent badges on this PR.

avg-api-times
avg-store-times
cpu-utilization
goroutines-in-use
heap-in-use
number-of-connected-devices
p99-api-times
p99-store-times
rps
stack-in-use

**Comment by matthewbirtch** [comment]
@koox00 I think the feature flag is off now on this test server so I can't test it

**Comment by jgilliam17** [comment]
@matthewbirtch I added it back, sorry about that - please refresh
I was verifying ack was not available without enterprise license

**Review by matthewbirtch — APPROVED** [review]
Approving this knowing that the remaining issues will be in follow-up bug fixes

**Comment by jgilliam17** [comment]
e2e-test MM_ENV="MM_FEATUREFLAGS_POSTPRIORITY=true"

**Comment by jgilliam17** [comment]
Thanks @koox00 
Tested, looks good to merge based on our discussion on follow up tickets.
Since this is behind feature flag, I am ok with checking the e2e report and fixing failures after merge.

**Review by jgilliam17 — APPROVED** [review]

**Comment by cwarnermm** [comment]
@koox00 - Is there any product or developer documentation updates needed for this PR beyond the API reference updates?

**Comment by cwarnermm** [comment]
@esethna - Does this PR require changes to the product or developer documentation?

**Comment by koox00** [comment]
> @koox00 - Is there any product or developer documentation updates needed for this PR beyond the API reference updates?

Hi @cwarnermm, sorry for the late reply, yes it will need product documentation

**Comment by esethna** [comment]
@cwarnermm the end user docs cover the functionality of acknowledgement, this is probably an update to the dev docs re APIS

**Comment by cwarnermm** [comment]
Thanks, @esethna. @koox00 - Are you open to creating a dev docs PR and/or an API reference PR that addresses the gaps in documentation? Happy to collaborate together to locate applicable docs content that needs review.

**Comment by cwarnermm** [comment]
@Willyfrog - Is a member of your team open to creating a dev docs PR and/or API reference PR for this Engineering PR?

**Comment by koox00** [comment]
> @Willyfrog - Is a member of your team open to creating a dev docs PR and/or API reference PR for this Engineering PR?

@cwarnermm we had created an API reference PR which was merged last week.
I can also create a dev docs PR ;)

Could you open a ticket at core 1's board and assign it to me?

**Comment by cwarnermm** [comment]
Thanks, @koox00! I've logged https://mattermost.atlassian.net/browse/MM-52269 and assigned it to you.

**Pull request merged**
